okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt

```diff
@@ -60,7 +63,9 @@ class ExchangeFinder(
   private val routeSelector: RouteSelector = RouteSelector(
       address, call.client.routeDatabase, call, eventListener)
   private var connectingConnection: RealConnection? = null
-  private var hasStreamFailure = false
+  private var refusedStreamCount = 0
```

Comment:
What do you think about pulling these out into a separate class.  Might help land it e.g. with a simple legacy strategy and the correct ones that you can unit test for?

Reply:
I think I'm skewed here because of frameworks I've used at scale, that were not simple binary healthy/unhealthy had more distinct abstractions here. For a mobile client where biggest change is likely network/route related, and server is most likely hidden behind wildcard host and a reverse proxy to individual servers, maybe this is fine.

Finagle https://github.com/twitter/finagle/blob/develop/finagle-core/src/main/scala/com/twitter/finagle/liveness/FailureAccrualPolicy.scala

RSocket https://github.com/rsocket/rsocket-java/blob/21c39bdaa60c6a916e3288048b268f2b756dd5e7/rsocket-core/src/main/java/io/rsocket/Availability.java

Reply:
I think for OkHttp we want to mostly track facts like stream failure counts, and let a user-defined class make the decisions.

Reply:
Conversation continues here: https://github.com/square/okhttp/pull/5787/files